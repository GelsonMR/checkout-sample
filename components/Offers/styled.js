import styled from 'styled-components'
import * as Typography from '../Typography'

export const Heading = styled(Typography.H4)`
  margin-bottom: 8px;
  color: ${props => props.theme.colors.black};
`

export const Form = styled.form`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-row-gap: 30px;
  grid-column-gap: 50px;

  > * {
    grid-column: 1 / 3;
  }
`

export const Email = styled(Typography.Footnote)`
  height: 24px;
  padding: 0 12px;
  display: inline-flex;
  margin-bottom: 30px;
  border-radius: 12px;
  align-items: center;
  color: ${props => props.theme.colors.black};
  border: 1px solid ${props => props.theme.colors.gray1};
`

export const Offer = styled(Typography.Footnote).attrs({ as: 'li' })`
  position: relative;
  display: grid;
  row-gap: 4px;
  padding: 20px;
  border: 1px solid ${props => props.theme.colors.primary};
  padding-right: 36px;
  border-radius: 15px;
  cursor: pointer;

  & + & {
    margin-top: 12px;
  }
`

export const Title = styled(Typography.Body2).attrs({ as: 'div' })`
  color: ${props => props.theme.colors.primary};
  font-weight: 700;
`

export const Caption = styled(Typography.Footnote)`
  display: flex;
  color: ${props => props.theme.colors.primary};
`

export const DiscountTag = styled(Typography.Footnote)`
  font-size: 10px;
  height: 16px;
  padding: 0 6px;
  display: flex;
  align-items: center;
  color: #FFF;
  background-color: ${props => props.theme.colors.secondary};
  margin-left: 12px;
  border-radius: 8px;
`

export const Installments = styled(Typography.Footnote)`
  font-size: 10px;
  color: ${props => props.theme.colors.secondary};
`

export const RadioButton = styled.div`
  position: absolute;
  top: 0;
  right: 20px;
  bottom: 0;
  width: 16px;
  height: 16px;
  margin: auto 0;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 2px solid ${props => props.theme.colors.gray3};
  border-radius: 8px;

  ${props => props.checked && `
    &::before {
      content: '';
      display: block;
      width: 8px;
      height: 8px;
      border-radius: 4px;
      background-color: ${props.theme.colors.primary};
    }
  `}
`

export const Info = styled(Typography.Footnote).attrs({ as: 'div' })`
  display: flex;
  align-items: center;
  justify-content: center;
  color: ${props => props.theme.colors.black};
  margin-top: 30px;
  cursor: default;
`

export const Icon = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 14px;
  height: 14px;
  margin-left: 12px;
`
