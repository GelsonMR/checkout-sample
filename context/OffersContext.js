import { createContext, useCallback, useContext, useEffect, useState } from 'react'
import { getOffers } from '../client'

export const OffersContext = createContext({})

export function OffersContextProvider({ children }) {
  const [isLoading, setIsLoading] = useState(false)
  const [offers, setOffers] = useState()
  const [selectedOffer, setSelectedOffer] = useState()
  const [error, setError] = useState(false)

  const fetchOffers = useCallback(async () => {
    setIsLoading(true)
    setError(false)
    try {
      const res = await getOffers()
      setOffers(res.data.sort((a, b) => a.order - b.order))
      setSelectedOffer(res.data[0])
    } catch (e) {
      console.error(e)
      setError(true)
    }
    setIsLoading(false)
  }, [])

  useEffect(() => {
    fetchOffers()
  }, [fetchOffers])

  const sharedState = {
    isLoading,
    offers,
    selectedOffer,
    error,
    setSelectedOffer
  }

  return (
    <OffersContext.Provider value={sharedState}>
      {children}
    </OffersContext.Provider>
  )
}

export function useOffers() {
  return useContext(OffersContext)
}
