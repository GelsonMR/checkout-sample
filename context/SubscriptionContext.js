import { createContext, useCallback, useContext, useState } from 'react'
import { subscribe } from '../client'

export const SubscriptionContext = createContext()

export function SubscriptionContextProvider({ children }) {
  const [isLoading, setIsLoading] = useState(false)
  const [subscription, setSubscription] = useState()
  const [error, setError] = useState(false)

  const fetchSubscribe = useCallback(async (params) => {
    setIsLoading(true)
    setError(false)
    try {
      const res = await subscribe(params)
      setSubscription({
        ...res.data,
        ...params
      })
      setIsLoading(false)
      return res
    } catch (e) {
      console.error(e)
      setError(true)
      setIsLoading(false)
      return { error: true }
    }
  }, [])

  const sharedState = {
    isLoading,
    subscription,
    error,
    subscribe: fetchSubscribe
  }

  return (
    <SubscriptionContext.Provider value={sharedState}>
      {children}
    </SubscriptionContext.Provider>
  )
}

export function useSubscription() {
  return useContext(SubscriptionContext)
}
