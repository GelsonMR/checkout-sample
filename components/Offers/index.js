import Image from 'next/image'
import { currency } from '../../utils'
import { useOffers } from '../../context/OffersContext'
import { useSubscription } from '../../context/SubscriptionContext'
import interrogationMarkCircle from '../../public/icons/interrogation-mark-circle.svg'
import { Caption, DiscountTag, Email, Heading, Icon, Info, Installments, Offer, RadioButton, Title } from './styled'
import Section from '../Section'

export default function Offers() {
  const { offers, selectedOffer, setSelectedOffer } = useOffers()
  const { isLoading } = useSubscription()

  return (
    <Section>
      <Heading>Confira seu plano:</Heading>
      <Email>fulano@cicrano.com.br</Email>
      {
        offers?.map((offer) => {
          const fullPriceLabel = currency.format(offer.fullPrice)
          const discountedPrice = offer.fullPrice - offer.discountAmmount
          const discountedPriceLabel = currency.format(discountedPrice)
          const discountedPerc = offer.discountPercentage * 100
          const monthlyPriceLabel = currency.format(discountedPrice / offer.installments)

          const handleOfferClick = () => {
            if (!isLoading) {
              setSelectedOffer(offer)
            }
          }

          return (
            <Offer key={offer.id} onClick={handleOfferClick}>
              <Title>{offer.title} | {offer.description}</Title>
              <Caption>
                De R$ {fullPriceLabel} | Por R$ {discountedPriceLabel}
                <DiscountTag>-{discountedPerc}%</DiscountTag>
              </Caption>
              <Installments>{offer.installments}x de R$ {monthlyPriceLabel}{offer.installments > 1 && '/mês'}</Installments>
              <RadioButton checked={offer.id === selectedOffer.id} />
            </Offer>
          )
        })
      }
      <Info>
        Sobre a cobrança
        <Icon>
          <Image src={interrogationMarkCircle} alt="Ajuda" />
        </Icon>
      </Info>
    </Section>
  )
}
