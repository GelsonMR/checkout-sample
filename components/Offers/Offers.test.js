import '@testing-library/jest-dom'
import { ThemeProvider } from 'styled-components'
import { render, screen, fireEvent, getByText } from '@testing-library/react'
import { OffersContext, OffersContextProvider } from '../../context/OffersContext'
import Offers from './index'

const mockTheme = { colors: {} }

it('renders offers', () => {
  const { container } = render(
    <ThemeProvider theme={mockTheme}>
      <Offers />
    </ThemeProvider>
  )
  expect(container).toMatchSnapshot()
})

const mockOffers =     [
  {
    "id":32,
    "storeId":"anual_parcelado_iugu",
    "title":"Premium Anual",
    "description":"Parcelado",
    "caption":"7 Dias Grátis",
    "fullPrice":600,
    "discountAmmount":60,
    "discountPercentage":0.1,
    "periodLabel":"mês",
    "period":"annually",
    "discountCouponCode":null,
    "order":1,
    "priority":1,
    "gateway":"iugu",
    "splittable":true,
    "installments":12,
    "acceptsCoupon":true
  },
  {
    "id":33,
    "storeId":"anual_a_vista_iugu",
    "title":"Premium Anual",
    "description":"À Vista",
    "caption":"7 Dias Grátis",
    "fullPrice":7200,
    "discountAmmount":720,
    "discountPercentage":0.1,
    "periodLabel":"ano",
    "period":"annually",
    "discountCouponCode":null,
    "order":2,
    "priority":2,
    "gateway":"iugu",
    "splittable":false,
    "installments":1,
    "acceptsCoupon":true
  }
]

it('renders home with 2 offers', async () => {
  render(
    <OffersContext.Provider value={{ offers: mockOffers, selectedOffer: mockOffers[0] }}>
      <ThemeProvider theme={mockTheme}>
        <Offers />
      </ThemeProvider>
    </OffersContext.Provider>
  )

  const container = document.querySelector('li')

  expect(container).toHaveLength(2)
})

it('changes selected offer', async () => {
  render(
    <OffersContext.Provider value={{ offers: mockOffers, selectedOffer: mockOffers[0] }}>
      <ThemeProvider theme={mockTheme}>
        <Offers />
      </ThemeProvider>
    </OffersContext.Provider>
  )
  
  const container = document.querySelector('li + li')
  
  fireEvent(
    container,
    new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }),
  )

  expect(container).toBe(1)
})
