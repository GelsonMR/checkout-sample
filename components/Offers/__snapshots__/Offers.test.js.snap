// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders offers 1`] = `
<div>
  <div
    class="sc-e5c0085b-0 hatemx"
  >
    <h4
      class="sc-2764c71a-0 sc-7df95e90-0 hiTdJi bKfnkt"
    >
      Confira seu plano:
    </h4>
    <span
      class="sc-2764c71a-1 sc-2764c71a-3 sc-7df95e90-2 gsTYYO dCDPex kxgKuG"
    >
      fulano@cicrano.com.br
    </span>
    <ul />
    <div
      class="sc-2764c71a-1 sc-2764c71a-3 sc-7df95e90-9 gsTYYO dCDPex hrZCZB"
    >
      Sobre a cobran√ßa
      <div
        class="sc-7df95e90-10 gsKZbp"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2724%27%20height=%2724%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="Ajuda"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
    </div>
  </div>
</div>
`;
